src/ring/middleware/gzip.clj

```diff
@@ -60,20 +60,31 @@
                       (dissoc "Content-Length")))
       (update-in [:body] piped-gzipped-input-stream)))
 
-(defn wrap-gzip [handler]
-  (fn [req]
-    (let [{:keys [body status] :as resp} (handler req)]
-      (if (and (= status 200)
-               (not (get-in resp [:headers "Content-Encoding"]))
-               (or
-                (and (string? body) (> (count body) 200))
-                (and (seq? body) @flushable-gzip?)
-                (instance? InputStream body)
-                (instance? File body)))
-        (let [accepts (get-in req [:headers "accept-encoding"] "")
-              match (re-find #"(gzip|\*)(;q=((0|1)(.\d+)?))?" accepts)]
-          (if (and match (not (contains? #{"0" "0.0" "0.00" "0.000"}
-                                         (match 3))))
-            (gzipped-response resp)
-            resp))
-        resp))))
+(defn- wrap-gzip* [req {:keys [body status] :as resp}]
+  (if (and (= status 200)
+           (not (get-in resp [:headers "Content-Encoding"]))
+           (or
+            (and (string? body) (> (count body) 200))
+            (and (seq? body) @flushable-gzip?)
+            (instance? InputStream body)
+            (instance? File body)))
+    (let [accepts (get-in req [:headers "accept-encoding"] "")
+          match (re-find #"(gzip|\*)(;q=((0|1)(.\d+)?))?" accepts)]
+      (if (and match (not (contains? #{"0" "0.0" "0.00" "0.000"}
+                                     (match 3))))
+        (gzipped-response resp)
+        resp))
+    resp))
+
+(defn wrap-gzip
+  "Ring middleware that GZIPs response if client can handle it."
+  [handler]
+  (fn
+    ([request]
+     (wrap-gzip* request (handler request)))
+
+    ([request respond raise]
+     (handler
+      request
+      (comp respond (partial wrap-gzip* request))
```

Comment:
Can you make this match the style at https://github.com/ring-clojure/ring-json/blob/0.5.0-beta1/src/ring/middleware/json.clj#L125? Composing multiple functions like this is not that easy to understand immediately.